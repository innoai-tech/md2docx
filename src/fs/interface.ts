import { type Value, VFile } from "vfile";

export interface FS {
  load(path: string, opt?: { cwd?: string }): Promise<VFile>;

  save(path: string, data: Value, opt?: { cwd?: string }): Promise<void>;
}
