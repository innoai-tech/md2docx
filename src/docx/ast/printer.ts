import { isBuffer, isNull, isUndefined, map, repeat } from "lodash";
import type { FileChild } from "./nodes.ts";

export const printAST = (children: FileChild[]): string => {
  const print = (root: any, depth = 0) => {
    const { children: _1, type: _2, data: _3, options } = root as any;

    const attrsStr = map(options, (v, k) => {
      if (isNull(v) || isUndefined(v)) {
        return "";
      }
      if (isBuffer(v)) {
        return `${k}=[Buffer]`;
      }

      return `${k}=${JSON.stringify(v)}`;
    })
      .filter((v) => v)
      .sort()
      .join(",");

    const childrenStr: string =
      root.children
        ?.map((c: any) => {
          return `${repeat("  ", depth + 1)}${print(c as any, depth + 1)}`;
        })
        .join("\n") ?? "";

    return `${root.type}${attrsStr ? `{${attrsStr}}` : ""}${
      childrenStr ? `\n${childrenStr}` : ""
    }`;
  };

  return children.map((c) => print(c)).join("\n");
};
