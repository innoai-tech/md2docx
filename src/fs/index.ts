export * from "./interface.ts";
export * from "./os.ts";
export * from "./mem.ts";
