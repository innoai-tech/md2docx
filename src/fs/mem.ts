import type { FS } from "./interface.ts";
import { type Value, VFile } from "vfile";
import { join } from "path";
import { relative } from "path";

export class MemFS implements FS {
  constructor(
    protected files: Record<string, string>,
    protected prefix = process.cwd(),
  ) {}

  async load(path: string, opt?: { cwd?: string }): Promise<VFile> {
    const cwd = opt?.cwd ?? ".";
    const p = relative(this.prefix, join(cwd, path));

    return new VFile({
      path: path,
      value: this.files[p],
      cwd: cwd,
    });
  }

  async save(path: string, data: Value, opt?: { cwd?: string }): Promise<void> {
    const cwd = opt?.cwd ?? ".";
    const p = relative(this.prefix, join(cwd, path));
    this.files[p] = String(data);
  }
}
