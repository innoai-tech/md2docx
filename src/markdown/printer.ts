import { type Root } from "mdast";
import { isBuffer, isNull, isUndefined, map, repeat } from "lodash";

export const printAST = (root: Root): string => {
  const print = (root: Root, depth = 0) => {
    const {
      position: _0,
      children: _1,
      type: _3,
      data,
      attributes,
      ...props
    } = root as any;

    const attrsStr = map({ ...data, ...props, ...attributes }, (v, k) => {
      if (isNull(v) || isUndefined(v)) {
        return "";
      }
      if (isBuffer(v)) {
        return `${k}=[Buffer]`;
      }

      return `${k}=${JSON.stringify(v)}`;
    })
      .filter((v) => v)
      .sort()
      .join(",");

    const childrenStr: string =
      root.children
        ?.map((c) => {
          return `${repeat("  ", depth + 1)}${print(c as any, depth + 1)}`;
        })
        .join("\n") ?? "";

    return `${root.type}${attrsStr ? `{${attrsStr}}` : ""}${
      childrenStr ? `\n${childrenStr}` : ""
    }`;
  };

  return print(root);
};
