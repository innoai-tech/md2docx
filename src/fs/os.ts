import type { FS } from "./interface.ts";
import { type Value, VFile } from "vfile";
import { readFile, mkdir, writeFile } from "fs/promises";
import { dirname, join } from "path";

export class OsFS implements FS {
  async load(path: string, opt?: { cwd: string }): Promise<VFile> {
    const cwd = opt?.cwd ?? ".";

    return new VFile({
      path: path,
      value: await readFile(join(cwd, path)),
      cwd: cwd
    });
  }

  async save(path: string, data: Value, opt?: { cwd?: string }): Promise<void> {
    const cwd = opt?.cwd ?? ".";
    const p = join(cwd, path);
    await mkdir(dirname(p), { recursive: true });
    await writeFile(p, data);
  }
}
