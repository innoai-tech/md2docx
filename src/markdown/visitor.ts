import type {
  Root,
  DefinitionContent,
  PhrasingContent,
  BlockContent,
  RootContent,
} from "mdast";
import type { VFile } from "vfile";
import { parse } from "./markdown.ts";
import { Cursor } from "./Cursor.ts";
import type { FS } from "../fs";
import { OsFS } from "../fs";

export type Exactly<Union, Type> = Union extends { type: Type } ? Union : never;

export type MdastNode =
  | RootContent
  | BlockContent
  | PhrasingContent
  | DefinitionContent;

export type Visitors = {
  [K in MdastNode["type"]]?: (
    c: Cursor<Exactly<MdastNode, K>>,
    ctx: VisitorContext,
  ) => Promise<void>;
};

export type VisitorContext = {
  fs: FS;
  parser: Parser;
  visitChildren(c: Cursor, options?: ProcessOptions): Promise<void>;
};

export type ProcessOptions = {
  fs?: FS;
  visitors?: Visitors[];
};

export interface Parser {
  parse: (f: VFile) => Cursor<Root>;
}

export const process = async (
  file: VFile,
  options: ProcessOptions,
): Promise<Root> => {
  const w = new Context(
    {
      parse: (file) => {
        return Cursor.of(parse(file.value), { file });
      },
    },
    options.fs ?? new OsFS(),
    options.visitors ?? [],
  );

  const root = w.parser.parse(file);
  await w.visitChildren(root);
  return root.node;
};

class Context implements VisitorContext {
  constructor(
    public parser: Parser,
    public fs: FS,
    protected visitors: Visitors[],
  ) {}

  async visitChildren(c: Cursor, options: ProcessOptions = {}): Promise<void> {
    for (const child of c.children()) {
      for (const v of [...this.visitors, ...(options.visitors ?? [])]) {
        const visitor = (v as any)[child.node.type];

        if (!visitor) {
          await this.visitChildren(child);
          continue;
        }

        await visitor(child, this);
      }
    }
  }
}
