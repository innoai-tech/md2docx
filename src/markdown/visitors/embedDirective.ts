import { type VisitorContext, type Visitors } from "../visitor.ts";
import { VFile } from "vfile";
import { isUndefined } from "lodash";
import { Cursor } from "../Cursor.ts";
import type { Root } from "mdast";

export interface Loader {
  test: RegExp;

  transform(file: VFile, ctx: VisitorContext): Promise<Cursor<Root>>;
}

export const defaultLoaders: Loader[] = [
  {
    test: /\.md$/,
    transform: async (file, ctx) => {
      return ctx.parser.parse(
        new VFile({
          path: file.path,
          cwd: file.cwd,
          value: file.value
        })
      );
    }
  }
];

const normalizeHeadingDepth = (headingDepthStarts: number): Visitors => {
  let delta: number;

  return {
    async heading(c, _) {
      if (isUndefined(delta)) {
        delta = headingDepthStarts - c.node.depth + 1;
      }
      c.node.depth += delta;
    }
  };
};

export const embedDirective = ({
                                 loaders = defaultLoaders
                               }: {
  loaders?: Loader[];
}): Visitors => {
  let headingDepth: number | null = null;

  return {
    async heading(c, ctx) {
      headingDepth = c.node.depth;
      await ctx.visitChildren(c);
    },

    async thematicBreak(c, ctx) {
      headingDepth = null;
      await ctx.visitChildren(c);
    },

    async leafDirective(c, ctx) {
      switch (c.node.name) {
        case "embed": {
          const url = c.node.attributes!["url"]!;
          const inHeading = c.node.attributes!["inHeading"];

          for (const loader of loaders) {
            if (loader.test.test(url)) {
              const subRoot: Cursor<Root> = await loader.transform(
                await ctx.fs.load(url, c.file),
                ctx
              );

              await ctx.visitChildren(subRoot, {
                visitors: [
                  normalizeHeadingDepth(inHeading ? parseInt(inHeading) : headingDepth ?? 0)
                ]
              });

              c.replaceWith(...subRoot.node.children);
            }
          }
        }
      }
    }
  };
};
