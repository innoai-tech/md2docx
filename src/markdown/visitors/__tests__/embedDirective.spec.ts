import { describe, expect, it } from "bun:test";
import { VFile } from "vfile";
import { process } from "../../visitor.ts";
import { printAST } from "../../printer.ts";
import { embedDirective } from "../embedDirective.ts";
import { MemFS } from "../../../fs";

describe("#embedDirective", () => {
  it("leafDirective should replaced with embeded", async () => {
    const node = await process(
      new VFile(
        `
# 一级标题

::embed{url=./sub.md}
`,
      ),
      {
        fs: new MemFS({
          "sub.md": `
# 标题

内容内容
`,
        }),
        visitors: [embedDirective({})],
      },
    );

    expect(printAST(node)).toMatchSnapshot();
  });
});
