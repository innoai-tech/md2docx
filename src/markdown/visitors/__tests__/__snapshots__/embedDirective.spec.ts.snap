// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`#embedDirective leafDirective should replaced with embeded 1`] = `
"root
  heading{depth=1}
    text{value="一级标题"}
  heading{depth=2}
    text{value="标题"}
  paragraph
    text{value="内容内容"}"
`;
